scalableminds/webknossos
# Bounding Box tool improvements

This PR keeps newly created bounding boxes via button interaction (and not created via drag) within the dataset bounding box to prevent initial invalid datasets (fixes #7353). Moreover, I added the option to move a bounding box around via ctrl +left mouse down. 

### URL of deployed dev instance (used for testing):
- https://bboxtoolimprovements.webknossos.xyz/

### Steps to test:
- Create an annotation of any dataset
- In the dataset move to the edge of the dataset. Then go to the bounding box tab and create new bounding boxes. The bounding boxes should be kept within the dataset bounds. Click in the bbox coordinate input box. It should not be marked red as invalid. 
- Move completely out of the dataset and hide the create bbox button again. Now the bounding box should appear as normal, completely out of the bounds of the dataset. Clicking focusing the coordinate input of this bbox should mark it red.
- Select the bounding box tool. Hold control and mouse drag to move a bounding box. Try this at different zoom levels. It should appear as expected. In very high zoom levels there might be minor differences between how much the mouse pointer and the bbox moves due to rounding issues. But I'd say that this is ok.

Edit: Nvm: the bbox coordinate is not marked red anymore as claimed by issue #7353. Maybe this was changed in the past.

### TODOs:
- [x] regarding bbox moving: The current version is not that much in sync with the amount actually dragged as well as the alt button is already being used to move within the dataset. => The UX doesn't feel smooth. Maybe use a different button (ctrl / meta) and save the initial mouse down position to calculate the actual delta movement
  - Now, the ctrl / meta key is used. This makes this a little less confusing but using ctrl / meta to move the bounding box is not as intuitive as using alt. Else we could disable moving in the bbox tool via alt and use alt as a key to trigger the move bounding box.
- [x] Think about how to properly handle: In case the user creates a bounding box completely outside the dataset. 
  - Just create the bounding box without restricting its bounds to the dataset and let the user cook.
   ![](https://i.kym-cdn.com/photos/images/newsfeed/002/488/659/18a.jpg)

### Issues:
- fixes #7353 and #7756

------
(Please delete unneeded items, merge only when none are left open)
